File: writing/draft-parsing-programming-languages/parsing.clj
;; expr = product eof
;;
;; product
;;  = additive ("*" additive)*
;;  | additive ("|" additive)*
;;
;; additive
;;  = primitive ("+" primitive)*
;;  | primitive ("-" primitive)*
;;
;; primitive
;;  = number
;;  | "(" expr ")"

;; Runtime

(defn p-eof [input]
  (if (= (count input) 0) [true nil input] [false nil input]))

(defn p-number [input]
  (cond
    (= (first input) \1) [true 1 (rest input)]
    (= (first input) \2) [true 2 (rest input)]
    (= (first input) \3) [true 3 (rest input)]
    (= (first input) \4) [true 4 (rest input)]
    (= (first input) \5) [true 5 (rest input)]
    (= (first input) \6) [true 6 (rest input)]
    (= (first input) \7) [true 7 (rest input)]
    (= (first input) \8) [true 8 (rest input)]
    (= (first input) \9) [true 9 (rest input)]
    (= (first input) \0) [true 0 (rest input)]
    :else [false nil input]))

(defn p-str [text]
  (fn [input]
    (cond
      (= (take (count text) input) (seq text))
      [true (take (count text) input) (drop (count text) input)]
      :else
      [false nil input])))

(defn p-and-helper [parsers output input]
  (if (> (count parsers) 0)
    (let [first-result ((first parsers) input)]
      (if (first first-result)
        (let [rest-result (p-and-helper (rest parsers)
                                        (conj output (nth first-result 1))
                                        (nth first-result 2))]
          (if (first rest-result)
            rest-result
            [false nil input]))
        [false nil input]))
    [true output input]))

(defn p-and [parsers]
  (fn [input]
    (p-and-helper parsers [] input)))

(defn p-or-helper [parsers input]
  (if (> (count parsers) 0)
    (let [first-result ((first parsers) input)]
      (if (first first-result)
        first-result
        (p-or-helper (rest parsers) input)))
    [false nil input]))

(defn p-or [parsers]
  (fn [input]
    (p-or-helper parsers input)))

(defn p-star-helper [parser result input]
  (let [first-result (parser input)]
    (if (first first-result)
      (p-star-helper parser (conj result (nth first-result 1)) (nth first-result 2))
      [true result input])))

(defn p-star [parser]
  (fn [input]
    (p-star-helper parser [] input)))

(defn run [parser input]
  (let [[success result input] (parser input)]
    (if success
      result
      nil)))

;; DSL

(declare expr
         product product-clause-1 product-clause-2
         additive additive-clause-1 additive-clause-2
         primitive primitive-clause-1 primitive-clause-2)

(defn expr [input]
  ((p-and [product p-eof]) input))

(defn product [input]
  ((p-or [product-clause-1 product-clause-2]) input))

(defn product-clause-1 [input]
  ((p-and [additive (p-star (p-and [(p-str "+") additive]))]) input))

(defn product-clause-2 [input]
  ((p-and [additive (p-star (p-and [(p-str "-") additive]))]) input))

(defn additive [input]
  ((p-or [additive-clause-1 additive-clause-2]) input))

(defn additive-clause-1 [input]
  ((p-and [primitive (p-star (p-and [(p-str "*") primitive]))]) input))

(defn additive-clause-2 [input]
  ((p-and [primitive (p-star (p-and [(p-str "|") primitive]))]) input))

(defn primitive [input]
  ((p-or [primitive-clause-1 primitive-clause-2]) input))

(defn primitive-clause-1 [input]
  (p-number input))

(defn primitive-clause-2 [input]
  ((p-and [(p-str "(") expr (p-str ")")]) input))

(run expr "3*4")
(run expr "3+4*5")

File: writing/draft-parsing-programming-languages/parsing.py
# expr = level1 eof
#
# level1
#  = level2 ("+" level2 { "add()\n" })*
#  | level2 ("-" level2 { "sub()\n" })*
#
# level2
#  = level3 ("*" level3 { "mult()\n" })*
#  | level3 ("/" level3 { "div()\n"  })*
#
# level3
#  = number { "push(" % ")\n" }
#  | "(" expr ")"


class MaybeParseError(Exception):
    pass


class BaseGrammar(object):

    def parse(self, text):
        self._input = text
        self._input_pos = 0
        self._output = []
        try:
            self.main()
            return "".join(self._output)
        except MaybeParseError as e:
            print("+++")
            print("".join(self._output))
            print("ERROR: {}".format(e))
            print("+++")

    def _number(self):
        self._choice([
            lambda: self._text("0"),
            lambda: self._text("1"),
            lambda: self._text("2"),
            lambda: self._text("3"),
            lambda: self._text("4"),
            lambda: self._text("5"),
            lambda: self._text("6"),
            lambda: self._text("7"),
            lambda: self._text("8"),
            lambda: self._text("9"),
        ])

    def _text(self, text):
        if self._input[self._input_pos:].startswith(text):
            self._input_pos += len(text)
            self._last_match = text
        else:
            raise MaybeParseError("No match for {}".format(text))

    def _write(self, text):
        self._output.append(text)

    def _write_last(self):
        self._write(self._last_match)

    def _star(self, parser):
        while True:
            try:
                savepoint = self._save()
                parser()
            except MaybeParseError:
                self._restore(savepoint)
                return

    def _eof(self):
        if len(self._input[self._input_pos:]) == 0:
            return
        else:
            raise MaybeParseError("Not at end.")

    def _choice(self, parsers):
        for parser in parsers:
            try:
                savepoint = self._save()
                parser()
                return
            except MaybeParseError:
                self._restore(savepoint)
        raise MaybeParseError("No choice matched")

    def _save(self):
        return self._input_pos, len(self._output)

    def _restore(self, savepoint):
        input_pos, output_len = savepoint
        self._input_pos = input_pos
        self._output = self._output[:output_len]


class Grammar(BaseGrammar):

    def main(self):
        self.expr()

    def expr(self):
        self.level1()
        self._eof()

    def level1(self):
        def level1_1():
            self.level2()
            def level1_1_1():
                self._text("+")
                self.level2()
                self._write("add()\n")
            self._star(level1_1_1)
        def level1_2():
            self.level2()
            def level1_2_1():
                self._text("-")
                self.level2()
                self._write("sub()\n")
            self._star(level1_2_1)
        self._choice([level1_1, level1_2])

    def level2(self):
        def level2_1():
            self.level3()
            def level2_1_1():
                self._text("*")
                self.level3()
                self._write("mult()\n")
            self._star(level2_1_1)
        def level2_2():
            self.level3()
            def level2_2_1():
                self._text("/")
                self.level3()
                self._write("div()\n")
            self._star(level2_2_1)
        self._choice([level2_1, level2_2])

    def level3(self):
        def level3_1():
            self._number()
            self._write("push(")
            self._write_last()
            self._write(")\n")
        def level3_2():
            self._text("(")
            self.expr()
            self._text(")")
        self._choice([level3_1, level3_2])


print(Grammar().parse("3+8*5"))
